scheduler/pkg/kafka/pipeline/client.go

```
@@ -155,78 +157,199 @@ func getSubsriberIp() (string, error) {
-			if err != nil {
-				logger.WithError(err).Errorf("Failed to delete pipeline %s", event.PipelineName)
-			}
-		}
+		processor.handleEvent(event)
 	}
-	logger.Infof("Closing connection to scheduler")
-	defer func() {
+}
+
+func (pc *PipelineSchedulerClient) cleanup(stream scheduler.Scheduler_SubscribePipelineStatusClient) {
+	pc.ready.Store(false)
+	pc.logger.Info("Closing connection to scheduler")
+	if stream != nil {
 		_ = stream.CloseSend()
-	}()
-	return nil
+	}
+}
+
+func (ep *EventProcessor) handleEvent(event *scheduler.PipelineStatusResponse) {
+	switch event.Operation {
+	case pb.PipelineStatusResponse_PipelineDelete:
+		ep.handleDeletePipeline(event)
+	case pb.PipelineStatusResponse_PipelineCreate:
+		ep.handleCreateOrUpdatePipeline(event)
+	}
+}
+
+type EventProcessor struct {
+	client         *PipelineSchedulerClient
+	grpcClient     scheduler.SchedulerClient
+	subscriberName string
+	logger         *logrus.Entry
+}
+
+func (ep *EventProcessor) handleDeletePipeline(event *scheduler.PipelineStatusResponse) {
+	pv, err := pipeline.CreatePipelineVersionWithStateFromProto(event.Versions[0])
+	if err != nil {
+		ep.reportFailure(
+			chainer.PipelineUpdateMessage_Delete,
+			nil,
+			fmt.Sprintf("No existing pipeline %s to delete", event.PipelineName),
+			event.Timestamp,
+			nil,
+		)
+		return
+	}
+
+	err = ep.client.pipelineInferer.DeletePipeline(event.PipelineName, false)
+	if err != nil {
+		ep.reportFailure(
+			chainer.PipelineUpdateMessage_Delete,
+			pv,
+			fmt.Sprintf("Failed to delete pipeline %s", event.PipelineName),
+			event.Timestamp,
+			err,
+		)
+		return
+	}
+
+	message := fmt.Sprintf("Pipeline %s deleted", event.PipelineName)
+	ep.reportSuccess(chainer.PipelineUpdateMessage_Delete, pv, message, event.Timestamp)
+}
+
+func (ep *EventProcessor) handleCreateOrUpdatePipeline(event *scheduler.PipelineStatusResponse) {
+	pv, err := pipeline.CreatePipelineVersionWithStateFromProto(event.Versions[0])
```

Could we still log a warning in case the event contains multiple versions? We don't want to silently ignore unexpected input.